File: ios-swift/DASHit/Models/Order.swift
import Foundation
import CoreLocation

public enum OrderStatus: String, Codable, CaseIterable {
    case placed = "placed"
    case packing = "packing"
    case outForDelivery = "out_for_delivery"
    case delivered = "delivered"
    case cancelled = "cancelled"
    
    /// Orders are also written by the web admin and driver consoles, which use
    /// free-form labels ("Packed", "On the way"). Map those instead of failing
    /// to decode the whole order.
    public init(from decoder: Decoder) throws {
        let raw = try decoder.singleValueContainer().decode(String.self)
        self = OrderStatus(rawValue: raw) ?? OrderStatus(stage: DeliveryStage(status: raw))
    }
    
    public init(stage: DeliveryStage) {
        switch stage {
        case .placed: self = .placed
        case .packing: self = .packing
        case .onTheWay: self = .outForDelivery
        case .delivered: self = .delivered
        case .cancelled: self = .cancelled
        }
    }
    
    public var stage: DeliveryStage { DeliveryStage(status: rawValue) }
    
    public var title: String {
        switch self {
        case .placed: return "Order Placed"
        case .packing: return "Packing at Store"
        case .outForDelivery: return "Out for Delivery"
        case .delivered: return "Delivered"
        case .cancelled: return "Order Cancelled"
        }
    }
    
    public var progress: Double { stage.progress }
    
    public var iconName: String {
        switch self {
        case .placed: return "checkmark.circle.fill"
        case .packing: return "shippingbox.fill"
        case .outForDelivery: return "scooter"
        case .delivered: return "house.fill"
        case .cancelled: return "xmark.circle.fill"
        }
    }
}

public struct DriverLiveTracking: Codable, Hashable {
    public let lat: Double
    public let lng: Double
    public let heading: Double?
    public let speed: Double?
    public let updatedAt: Double?
    
    public var coordinate: CLLocationCoordinate2D {
        CLLocationCoordinate2D(latitude: lat, longitude: lng)
    }
}

public struct Order: Codable, Identifiable, Hashable {
    public let id: String
    public let userId: String
    public let items: [CartItem]
    public let subtotal: Double
    public let deliveryFee: Double
    public let discount: Double
    public let grandTotal: Double
    public var status: OrderStatus
    public let createdAt: Double
    public let deliveryAddress: DeliveryAddress
    public let paymentMethod: String // "cod", "apple_pay"
    public let paymentStatus: String // "pending", "completed"
    public var driverId: String?
    public var driverName: String?
    public var driverPhone: String?
    public var etaMinutes: Int?
    public var tracking: DriverLiveTracking?
    
    public init(
        id: String,
        userId: String,
        items: [CartItem],
        subtotal: Double,
        deliveryFee: Double,
        discount: Double,
        grandTotal: Double,
        status: OrderStatus = .placed,
        createdAt: Double = Date().timeIntervalSince1970,
        deliveryAddress: DeliveryAddress,
        paymentMethod: String = "cod",
        paymentStatus: String = "pending",
        driverId: String? = nil,
        driverName: String? = nil,
        driverPhone: String? = nil,
        etaMinutes: Int? = 8,
        tracking: DriverLiveTracking? = nil
    ) {
        self.id = id
        self.userId = userId
        self.items = items
        self.subtotal = subtotal
        self.deliveryFee = deliveryFee
        self.discount = discount
        self.grandTotal = grandTotal
        self.status = status
        self.createdAt = createdAt
        self.deliveryAddress = deliveryAddress
        self.paymentMethod = paymentMethod
        self.paymentStatus = paymentStatus
        self.driverId = driverId
        self.driverName = driverName
        self.driverPhone = driverPhone
        self.etaMinutes = etaMinutes
        self.tracking = tracking
    }
}

File: android-compose/app/src/main/kotlin/com/dashit/app/data/model/Order.kt
package com.dashit.app.data.model

import kotlin.math.ceil
import kotlin.math.max
import kotlin.random.Random

/**
 * Where an order is, normalised from whatever status string the admin, driver,
 * web or iOS app wrote. Same stages, progress floors and copy as the iOS app's
 * `DeliveryStage`, so both apps describe an order the same way.
 */
enum class OrderStatus(val title: String, val rawValue: String) {
    PLACED("Order Placed", "Placed"),
    PACKING("Packing at Store", "Packing"),
    OUT_FOR_DELIVERY("Out for Delivery", "Out for Delivery"),
    DELIVERED("Delivered", "Delivered"),
    CANCELLED("Order Cancelled", "Cancelled");

    /** The web tracker's ORDER_PROGRESS_FLOOR: 12 / 34 / 58 / 100. */
    val floor: Double
        get() = when (this) {
            PLACED -> 0.12
            PACKING -> 0.34
            OUT_FOR_DELIVERY -> 0.58
            DELIVERED -> 1.0
            CANCELLED -> 0.0
        }

    val isFinished: Boolean get() = this == DELIVERED || this == CANCELLED

    /** Stage floor, raised by the rider's live 0–100 progress while riding. */
    fun progress(live: Double?): Double {
        if (this != OUT_FOR_DELIVERY || live == null || live <= 0) return floor
        return minOf(0.97, max(floor, live / 100))
    }

    fun headline(riderName: String?): String = when (this) {
        PLACED -> "Order placed"
        PACKING -> "Packing your order"
        OUT_FOR_DELIVERY -> riderName?.trim()?.takeIf { it.isNotEmpty() }?.let { "$it is on the way" } ?: "On the way to you"
        DELIVERED -> "Order delivered"
        CANCELLED -> "Order cancelled"
    }

    fun subtitle(etaMinutes: Int, itemCount: Int): String = when (this) {
        PLACED -> "Hub is picking fresh items"
        PACKING -> if (itemCount > 0) "Packing $itemCount item${if (itemCount == 1) "" else "s"} at the hub" else "Items are packed & sealed"
        OUT_FOR_DELIVERY -> if (etaMinutes <= 1) "Arriving now" else "Arriving in $etaMinutes mins"
        DELIVERED -> "Handed over safely"
        CANCELLED -> "This order has been cancelled"
    }

    companion object {
        /** "out" / "way" are checked before "deliver" so "out_for_delivery" is not read as delivered. */
        fun fromString(value: String): OrderStatus {
            val s = value.lowercase()
            return when {
                "cancel" in s -> CANCELLED
                listOf("out", "way", "rider", "dispatch", "transit").any { it in s } -> OUT_FOR_DELIVERY
                listOf("deliver", "arrived", "complete").any { it in s } -> DELIVERED
                listOf("pack", "bag", "ready").any { it in s } -> PACKING
                else -> PLACED
            }
        }
    }
}

/** The rider's live telemetry, `orders/{id}/tracking/live`, written by the driver app. */
data class DriverLiveTracking(
    val lat: Double,
    val lng: Double,
    val heading: Double? = null,
    val speed: Double? = null,
    val etaMinutes: Int? = null,
    val distanceFormatted: String? = null,
    val statusText: String? = null,
    /** 0–100, as the driver app writes it. */
    val progress: Double? = null,
    val stopsAhead: Int? = null
)

data class Order(
    val id: String,
    val userId: String,
    val items: List<CartItem>,
    val subtotal: Double,
    val deliveryFee: Double,
    val discount: Double,
    val grandTotal: Double,
    val status: OrderStatus = OrderStatus.PLACED,
    /** Epoch milliseconds. */
    val createdAt: Long = System.currentTimeMillis(),
    val deliveryAddress: DeliveryAddress = DeliveryAddress(),
    val paymentMethod: String = "Cash on Delivery",
    val paymentStatus: String = "pending",
    val driverId: String? = null,
    val driverName: String? = null,
    val driverPhone: String? = null,
    val etaMinutes: Int? = 8,
    val tracking: DriverLiveTracking? = null,
    /** The 4-digit code the rider asks for at the door. */
    val otp: String? = null,
    val couponCode: String? = null,
    /** Epoch milliseconds; set on replacement orders so the window keeps its deadline. */
    val modifyWindowEndsAt: Long? = null,
    val replacesOrderId: String? = null
) {
    val itemCount: Int get() = items.sumOf { it.qty }

    val modifyWindowEnd: Long get() = modifyWindowEndsAt ?: (createdAt + MODIFY_WINDOW_MS)

    /** Seconds left to add items or cancel: only while the store hasn't started. */
    fun modifySecondsRemaining(now: Long = System.currentTimeMillis()): Int {
        if (status != OrderStatus.PLACED) return 0
        return max(0, ceil((modifyWindowEnd - now) / 1000.0).toInt())
    }

    companion object {
        const val MODIFY_WINDOW_MS = 60_000L

        /** "DSH-" + clock digits + random digits, the same shape as the iOS order codes. */
        fun newCode(): String {
            val clock = (System.currentTimeMillis() / 1000) % 10_000
            return "DSH-%04d%d".format(clock, Random.nextInt(1000, 10000))
        }

        fun newDeliveryCode(): String = Random.nextInt(1000, 10000).toString()
    }
}
